<a href="https://colab.research.google.com/github/haliberry29/SadiyaTijjaniBaba_MachineLearning/blob/main/AI_Powered_Agribusiness_Pricing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

| Column Name                     | Description                                                                      |
| ------------------------------- | -------------------------------------------------------------------------------- |
| `farm_id`                       | Unique identifier for each farm operation or agricultural site                   |
| `region`                        | Geographic region where the farm is located                                      |
| `crop_type`                     | Type of crop being cultivated (e.g., corn, soybean, wheat)                       |
| `soil_moisture_%`               | Soil water content percentage indicating moisture availability                   |
| `soil_pH`                       | Soil acidity/alkalinity measurement affecting crop growth                        |
| `temperature_C`                 | Average environmental temperature in Celsius                                     |
| `rainfall_mm`                   | Total rainfall amount in millimeters                                             |
| `humidity_%`                    | Atmospheric humidity percentage                                                  |
| `sunlight_hours`                | Number of sunlight exposure hours received                                       |
| `irrigation_type`               | Irrigation method used (drip, sprinkler, flood, etc.)                            |
| `fertilizer_type`               | Type of fertilizer applied to crops                                              |
| `pesticide_usage_ml`            | Amount of pesticide applied in milliliters                                       |
| `sowing_date`                   | Date when crops were planted                                                     |
| `harvest_date`                  | Date when crops were harvested                                                   |
| `total_days`                    | Total crop growth duration between sowing and harvest                            |
| `yield_kg_per_hectare`          | Crop production yield per hectare                                                |
| `sensor_id`                     | Identifier for IoT or agricultural monitoring sensor                             |
| `timestamp`                     | Date-time record of observation or sensor reading                                |
| `latitude`                      | Geographic latitude coordinate of the farm                                       |
| `longitude`                     | Geographic longitude coordinate of the farm                                      |
| `NDVI_index`                    | Normalized Difference Vegetation Index measuring crop health from remote sensing |
| `crop_disease_status`           | Indicates whether crops are healthy or affected by disease                       |
| `product_price`                 | Selling price of agricultural product or input                                   |
| `quantity_sold`                 | Number of units sold                                                             |
| `revenue`                       | Total revenue generated from sales                                               |
| `discount_pct`                  | Discount percentage applied during sale                                          |
| `corn_futures_price`            | Market futures price for corn commodity                                          |
| `soybean_futures_price`         | Market futures price for soybean commodity                                       |
| `fertilizer_market_index`       | Market index representing fertilizer price trends                                |
| `diesel_price`                  | Diesel fuel price impacting machinery and logistics costs                        |
| `inflation_rate`                | General inflation percentage affecting operational costs                         |
| `competitor_price`              | Estimated competitor pricing for similar products                                |
| `customer_segment`              | Classification of customer type (small farm, enterprise, cooperative, etc.)      |
| `farm_size_hectares`            | Total farm operational area in hectares                                          |
| `purchase_frequency`            | Number of purchases made within a defined period                                 |
| `customer_lifetime_value`       | Estimated long-term customer revenue value                                       |
| `demand_index`                  | Relative measure of market demand intensity                                      |
| `regional_market_share_pct`     | Estimated percentage of regional market controlled                               |
| `week`                          | Week number extracted from timestamp                                             |
| `quarter`                       | Fiscal quarter extracted from timestamp                                          |
| `season`                        | Agricultural season classification (Spring, Summer, Fall, Winter)                |
| `holiday_flag`                  | Binary indicator identifying holiday periods                                     |
| `drought_index`                 | Numerical indicator of drought severity                                          |
| `heat_stress_index`             | Crop heat stress severity indicator                                              |
| `flood_risk_score`              | Estimated flood vulnerability score                                              |
| `soil_health_score`             | Composite soil quality and fertility score                                       |
| `crop_disease_risk`             | Estimated probability/risk category of crop disease occurrence                   |
| `water_availability_index`      | Relative water resource availability indicator                                   |
| `inventory_level`               | Available inventory stock quantity                                               |
| `supplier_risk_score`           | Supplier reliability and supply chain risk indicator                             |
| `logistics_cost`                | Transportation and delivery cost estimate                                        |
| `lead_time_days`                | Estimated delivery or replenishment lead time                                    |
| `unit_cost`                     | Estimated operational or production cost per unit                                |
| `profit_margin`                 | Profit percentage after subtracting costs from price                             |
| `pricing_recommendation`        | AI-generated pricing action recommendation                                       |
| `pricing_recommendation_reason` | Explanation supporting the pricing recommendation                                |
| `hormuz_scenario`               | Simulated geopolitical disruption scenario related to Strait of Hormuz           |
| `oil_price_shock_multiplier`    | Multiplier representing oil price shock severity                                 |
| `adjusted_diesel_price`         | Diesel price adjusted for geopolitical disruptions                               |
| `adjusted_fertilizer_index`     | Fertilizer price index adjusted for energy market shocks                         |
| `adjusted_logistics_cost`       | Logistics cost adjusted for fuel and supply chain disruption                     |
| `supply_chain_disruption_score` | Composite indicator of supply chain instability                                  |
| `farm_input_inflation_pct`      | Estimated inflation percentage in agricultural input costs                       |
| `adjusted_unit_cost`            | Production cost adjusted for geopolitical and inflationary pressure              |
| `revenue_risk_score`            | Estimated financial risk exposure to revenue                                     |
| `margin_compression_pct`        | Percentage reduction in profit margin due to rising costs                        |
| `adjusted_profit_margin`        | Profit margin after geopolitical and inflation adjustments                       |
| `revenue_at_risk`               | Estimated revenue potentially lost under disruption scenarios                    |
| `geopolitical_pricing_strategy` | Recommended pricing strategy under geopolitical conditions                       |
| `geopolitical_business_summary` | AI-generated business interpretation of geopolitical risk conditions             |


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [ ]:
df = pd.read_csv('/content/sample_data/Smart_Farming_Crop_Yield_2024_2.csv')

In [ ]:
df.columns

Index(['Unnamed: 0', 'farm_id', 'region', 'crop_type', 'soil_moisture_%',
       'soil_pH', 'temperature_C', 'rainfall_mm', 'humidity_%',
       'sunlight_hours', 'irrigation_type', 'fertilizer_type',
       'pesticide_usage_ml', 'sowing_date', 'harvest_date', 'total_days',
       'yield_kg_per_hectare', 'sensor_id', 'timestamp', 'latitude',
       'longitude', 'NDVI_index', 'crop_disease_status', 'product_price',
       'quantity_sold', 'revenue', 'discount_pct'],
      dtype='object')

In [ ]:
df.isnull().sum() #checking for empty/missing data

,0
Unnamed: 0,0
farm_id,0
region,0
crop_type,0
soil_moisture_%,0
soil_pH,0
temperature_C,0
rainfall_mm,0
humidity_%,0
sunlight_hours,0


#Analysis
two columns have missing data: irrigation_type, crop_disease_status

#Data Engineering

In [ ]:
df["product_price"] = np.random.normal(420, 40, len(df))

df["quantity_sold"] = (
    df["yield_kg_per_hectare"] * np.random.uniform(0.8, 1.3, len(df))
).astype(int)

df["revenue"] = df["product_price"] * df["quantity_sold"]

df["discount_pct"] = np.random.uniform(0, 0.15, len(df)) #percentage discount during sales

In [ ]:
import numpy as np
import pandas as pd

# -----------------------------
# RANDOM SEED FOR REPRODUCIBILITY
# -----------------------------
np.random.seed(42)

# Assume your dataframe already exists
# df = pd.read_csv("your_dataset.csv")

# -----------------------------
# MARKET & ECONOMIC FEATURES
# -----------------------------

# Corn futures price ($/bushel)
df["corn_futures_price"] = np.random.normal(6.2, 0.7, len(df)).round(2)

# Soybean futures price ($/bushel)
df["soybean_futures_price"] = np.random.normal(13.5, 1.2, len(df)).round(2)

# Fertilizer market index
df["fertilizer_market_index"] = np.random.normal(110, 15, len(df)).round(2)

# Diesel price ($/gallon)
df["diesel_price"] = np.random.normal(4.1, 0.4, len(df)).round(2)

# Inflation rate (%)
df["inflation_rate"] = np.random.uniform(2.0, 8.0, len(df)).round(2)

# Competitor product price
df["competitor_price"] = (
    df["product_price"] * np.random.uniform(0.9, 1.1, len(df))
).round(2)

# -----------------------------
# CUSTOMER & DEMAND FEATURES
# -----------------------------

# Customer segment
customer_segments = ["Small Farm", "Commercial", "Enterprise", "Cooperative"]

df["customer_segment"] = np.random.choice(
    customer_segments,
    len(df)
)

# Farm size (hectares)
df["farm_size_hectares"] = np.random.randint(10, 5000, len(df))

# Purchase frequency per year
df["purchase_frequency"] = np.random.randint(1, 15, len(df))

# Customer lifetime value ($)
df["customer_lifetime_value"] = np.random.normal(
    25000,
    8000,
    len(df)
).round(2)

# Demand index
df["demand_index"] = np.random.uniform(0.5, 1.5, len(df)).round(2)

# Regional market share (%)
df["regional_market_share_pct"] = np.random.uniform(
    1,
    35,
    len(df)
).round(2) #the estimated percentage of the regional market controlled by a farm, supplier, or agribusiness entity.
##| Value | Meaning                    |
##| ----- | -------------------------- |
##| 2%    | very small market presence |
##| 10%   | moderate market share      |
##| 25%   | dominant regional supplier |
##| 35%   | major market leader        |##


# -----------------------------
# TEMPORAL FEATURES
# -----------------------------

# Convert dates to datetime
df["sowing_date"] = pd.to_datetime(df["sowing_date"])
df["harvest_date"] = pd.to_datetime(df["harvest_date"])
df['timestamp'] = pd.to_datetime(df['timestamp'])
# Extract week
df["week"] = df["timestamp"].dt.isocalendar().week

# Extract quarter
df["quarter"] = df["timestamp"].dt.quarter

# Season mapping
def get_season(month):
    if month in [12, 1, 2]:
        return "Winter"
    elif month in [3, 4, 5]:
        return "Spring"
    elif month in [6, 7, 8]:
        return "Summer"
    else:
        return "Fall"

df["season"] = df["timestamp"].dt.month.apply(get_season)

# Holiday flag
df["holiday_flag"] = np.random.choice(
    [0, 1],
    len(df),
    p=[0.9, 0.1]
)

# -----------------------------
# ENVIRONMENTAL RISK FEATURES
# -----------------------------

# Drought index
df["drought_index"] = np.random.uniform(0, 5, len(df)).round(2)

# Heat stress index
df["heat_stress_index"] = (
    (df["temperature_C"] / 40) * 100
).round(2)

# Flood risk score
df["flood_risk_score"] = np.random.uniform(
    0,
    100,
    len(df)
).round(2)

# Soil health score
df["soil_health_score"] = np.random.uniform(
    50,
    100,
    len(df)
).round(2)

# Crop disease risk
disease_risk = ["Low", "Moderate", "High"]

df["crop_disease_risk"] = np.random.choice(
    disease_risk,
    len(df),
    p=[0.6, 0.3, 0.1]
)

# Water availability index
df["water_availability_index"] = np.random.uniform(
    0,
    1,
    len(df)
).round(2)

# -----------------------------
# INVENTORY & SUPPLY CHAIN
# -----------------------------

# Inventory level
df["inventory_level"] = np.random.randint(
    100,
    5000,
    len(df)
)

# Supplier risk score
df["supplier_risk_score"] = np.random.uniform(
    0,
    100,
    len(df)
).round(2)

##..delayed shipments
#geopolitical exposure
#higher procurement costs
##| Score  | Meaning                     |
##| ------ | --------------------------- |
##| 0–20   | Very low supplier risk      |
#| 20–50  | Moderate supplier risk      |
#| 50–80  | Elevated supplier risk      |
#| 80–100 | Severe supplier instability |

# Logistics cost ($)
df["logistics_cost"] = np.random.normal(
    500,
    120,
    len(df)
).round(2)

# Lead time (days): the number of days required for goods, supplies, or inventory to arrive after an order is placed.
df["lead_time_days"] = np.random.randint(
    1,
    30,
    len(df))
#reduce sales
#create stockouts
#hurt customer retention

# -----------------------------
# PROFITABILITY FEATURES
# -----------------------------

# Unit production cost
df["unit_cost"] = (df["product_price"] * np.random.uniform(0.5, 0.85, len(df))
).round(2)

# Profit margin: #the percentage of profit earned after subtracting production or operational costs from the selling price.
df["profit_margin"] = ((df["product_price"] - df["unit_cost"]) / df["product_price"]
    * 100).round(2) #i.e percentage of production cost

# -----------------------------
# AI PRICING RECOMMENDATION
# -----------------------------

def pricing_recommendation(row):

    if row["drought_index"] > 3 and row["demand_index"] > 1.1:
        return "Increase price by 5%"

    elif row["inventory_level"] > 4000:
        return "Offer 10% discount"

    elif row["competitor_price"] < row["product_price"]:
        return "Reduce price by 3%"

    else:
        return "Maintain current pricing"

df["pricing_recommendation"] = df.apply(
    pricing_recommendation,
    axis=1
)

# -----------------------------
# AI BUSINESS EXPLANATION
# -----------------------------

def recommendation_reason(row):

    if row["drought_index"] > 3:
        return "High drought risk increasing regional demand pressure."

    elif row["inventory_level"] > 4000:
        return "Excess inventory detected."

    elif row["competitor_price"] < row["product_price"]:
        return "Competitor pricing pressure identified."

    else:
        return "Market conditions stable."

df["pricing_recommendation_reason"] = df.apply(
    recommendation_reason,
    axis=1
)

# -----------------------------
# PREVIEW DATASET
# -----------------------------

print(df.head())

# -----------------------------
# SAVE ENHANCED DATASET
# -----------------------------

df.to_csv(
    "enhanced_agribusiness_pricing_dataset.csv",
    index=False
)

print("Enhanced dataset saved successfully.")

   Unnamed: 0   farm_id       region crop_type  soil_moisture_%  soil_pH  \
0           0  FARM0001  North India     Wheat            35.95     5.99   
1           1  FARM0002    South USA   Soybean            19.74     7.24   
2           2  FARM0003    South USA     Wheat            29.32     7.16   
3           3  FARM0004  Central USA     Maize            17.33     6.03   
4           4  FARM0005  Central USA    Cotton            19.37     5.92   

   temperature_C  rainfall_mm  humidity_%  sunlight_hours  ...  \
0          17.79        75.62       77.03            7.27  ...   
1          30.18        89.91       61.13            5.67  ...   
2          27.37       265.43       68.87            8.23  ...   
3          33.73       212.01       70.46            5.03  ...   
4          33.86       269.09       55.73            7.93  ...   

  crop_disease_risk water_availability_index  inventory_level  \
0               Low                     0.14              476   
1               

In [ ]:
import numpy as np
import pandas as pd

# -----------------------------
# RANDOM SEED FOR REPRODUCIBILITY
# -----------------------------
np.random.seed(42)

# Assume your dataframe already exists
# df = pd.read_csv("your_dataset.csv")

# -----------------------------
# MARKET & ECONOMIC FEATURES
# -----------------------------

# Corn futures price ($/bushel)
df["corn_futures_price"] = np.random.normal(6.2, 0.7, len(df)).round(2)

# Soybean futures price ($/bushel)
df["soybean_futures_price"] = np.random.normal(13.5, 1.2, len(df)).round(2)

# Fertilizer market index
df["fertilizer_market_index"] = np.random.normal(110, 15, len(df)).round(2)

# Diesel price ($/gallon)
df["diesel_price"] = np.random.normal(4.1, 0.4, len(df)).round(2)

# Inflation rate (%)
df["inflation_rate"] = np.random.uniform(2.0, 8.0, len(df)).round(2)

# Competitor product price
df["competitor_price"] = (
    df["product_price"] * np.random.uniform(0.9, 1.1, len(df))
).round(2)

# -----------------------------
# CUSTOMER & DEMAND FEATURES
# -----------------------------

# Customer segment
customer_segments = ["Small Farm", "Commercial", "Enterprise", "Cooperative"]

df["customer_segment"] = np.random.choice(
    customer_segments,
    len(df)
)

# Farm size (hectares)
df["farm_size_hectares"] = np.random.randint(10, 5000, len(df))

# Purchase frequency per year
df["purchase_frequency"] = np.random.randint(1, 15, len(df))

# Customer lifetime value ($)
df["customer_lifetime_value"] = np.random.normal(
    25000,
    8000,
    len(df)
).round(2)

# Demand index
df["demand_index"] = np.random.uniform(0.5, 1.5, len(df)).round(2)

# Regional market share (%)
df["regional_market_share_pct"] = np.random.uniform(
    1,
    35,
    len(df)
).round(2)

# -----------------------------
# TEMPORAL FEATURES
# -----------------------------

# Convert dates to datetime
df["sowing_date"] = pd.to_datetime(df["sowing_date"])
df["harvest_date"] = pd.to_datetime(df["harvest_date"])

# Extract week
df["week"] = df["timestamp"].dt.isocalendar().week

# Extract quarter
df["quarter"] = df["timestamp"].dt.quarter

# Season mapping
def get_season(month):
    if month in [12, 1, 2]:
        return "Winter"
    elif month in [3, 4, 5]:
        return "Spring"
    elif month in [6, 7, 8]:
        return "Summer"
    else:
        return "Fall"

df["season"] = df["timestamp"].dt.month.apply(get_season)

# Holiday flag
df["holiday_flag"] = np.random.choice(
    [0, 1],
    len(df),
    p=[0.9, 0.1]
)

# -----------------------------
# ENVIRONMENTAL RISK FEATURES
# -----------------------------

# Drought index
df["drought_index"] = np.random.uniform(0, 5, len(df)).round(2)

# Heat stress index
df["heat_stress_index"] = (
    (df["temperature_C"] / 40) * 100
).round(2)

# Flood risk score
df["flood_risk_score"] = np.random.uniform(
    0,
    100,
    len(df)
).round(2)

# Soil health score
df["soil_health_score"] = np.random.uniform(
    50,
    100,
    len(df)
).round(2)

# Crop disease risk
disease_risk = ["Low", "Moderate", "High"]

df["crop_disease_risk"] = np.random.choice(
    disease_risk,
    len(df),
    p=[0.6, 0.3, 0.1]
)

# Water availability index
df["water_availability_index"] = np.random.uniform(
    0,
    1,
    len(df)
).round(2)

# -----------------------------
# INVENTORY & SUPPLY CHAIN
# -----------------------------

# Inventory level
df["inventory_level"] = np.random.randint(
    100,
    5000,
    len(df)
)

# Supplier risk score
df["supplier_risk_score"] = np.random.uniform(
    0,
    100,
    len(df)
).round(2)

# Logistics cost ($)
df["logistics_cost"] = np.random.normal(
    500,
    120,
    len(df)
).round(2)

# Lead time (days)
df["lead_time_days"] = np.random.randint(
    1,
    30,
    len(df)
)

# -----------------------------
# PROFITABILITY FEATURES
# -----------------------------

# Unit production cost
df["unit_cost"] = (
    df["product_price"] * np.random.uniform(0.5, 0.85, len(df))
).round(2)

# Profit margin
df["profit_margin"] = (
    (df["product_price"] - df["unit_cost"])
    / df["product_price"]
    * 100
).round(2)

# -----------------------------
# AI PRICING RECOMMENDATION
# -----------------------------

def pricing_recommendation(row):

    if row["drought_index"] > 3 and row["demand_index"] > 1.1:
        return "Increase price by 5%"

    elif row["inventory_level"] > 4000:
        return "Offer 10% discount"

    elif row["competitor_price"] < row["product_price"]:
        return "Reduce price by 3%"

    else:
        return "Maintain current pricing"

df["pricing_recommendation"] = df.apply(
    pricing_recommendation,
    axis=1
)

# -----------------------------
# AI BUSINESS EXPLANATION
# -----------------------------

def recommendation_reason(row):

    if row["drought_index"] > 3:
        return "High drought risk increasing regional demand pressure."

    elif row["inventory_level"] > 4000:
        return "Excess inventory detected."

    elif row["competitor_price"] < row["product_price"]:
        return "Competitor pricing pressure identified."

    else:
        return "Market conditions stable."

df["pricing_recommendation_reason"] = df.apply(
    recommendation_reason,
    axis=1
)

# -----------------------------
# PREVIEW DATASET
# -----------------------------

print(df.head())

# -----------------------------
# SAVE ENHANCED DATASET
# -----------------------------

df.to_csv(
    "enhanced_agribusiness_pricing_dataset.csv",
    index=False
)

print("Enhanced dataset saved successfully.")

   Unnamed: 0   farm_id       region crop_type  soil_moisture_%  soil_pH  \
0           0  FARM0001  North India     Wheat            35.95     5.99   
1           1  FARM0002    South USA   Soybean            19.74     7.24   
2           2  FARM0003    South USA     Wheat            29.32     7.16   
3           3  FARM0004  Central USA     Maize            17.33     6.03   
4           4  FARM0005  Central USA    Cotton            19.37     5.92   

   temperature_C  rainfall_mm  humidity_%  sunlight_hours  ...  \
0          17.79        75.62       77.03            7.27  ...   
1          30.18        89.91       61.13            5.67  ...   
2          27.37       265.43       68.87            8.23  ...   
3          33.73       212.01       70.46            5.03  ...   
4          33.86       269.09       55.73            7.93  ...   

  crop_disease_risk water_availability_index  inventory_level  \
0               Low                     0.14              476   
1               

In [ ]:
df.crop_type.unique()

array(['Wheat', 'Soybean', 'Maize', 'Cotton', 'Rice'], dtype=object)

crop types in the dataset

In [ ]:
df.region.unique()

array(['North India', 'South USA', 'Central USA', 'East Africa',
       'South India'], dtype=object)

regions represented in the dataset

##geopolitical instability representation: strait of Hormuz

In [12]:
hormuz_scenarios = [
    "normal",
    "moderate_disruption",
    "severe_disruption",
    "closure"
]

df["hormuz_scenario"] = np.random.choice(
    hormuz_scenarios,
    len(df),
    p=[0.65, 0.20, 0.10, 0.05]
)

In [13]:
df.head(3)

,Unnamed: 0,farm_id,region,crop_type,soil_moisture_%,soil_pH,temperature_C,rainfall_mm,humidity_%,sunlight_hours,...,water_availability_index,inventory_level,supplier_risk_score,logistics_cost,lead_time_days,unit_cost,profit_margin,pricing_recommendation,pricing_recommendation_reason,hormuz_scenario
0,0,FARM0001,North India,Wheat,35.95,5.99,17.79,75.62,77.03,7.27,...,0.14,476,69.82,504.81,7,304.49,17.69,Reduce price by 3%,Competitor pricing pressure identified.,normal
1,1,FARM0002,South USA,Soybean,19.74,7.24,30.18,89.91,61.13,5.67,...,0.78,1072,62.38,248.98,28,365.76,21.82,Reduce price by 3%,High drought risk increasing regional demand p...,severe_disruption
2,2,FARM0003,South USA,Wheat,29.32,7.16,27.37,265.43,68.87,8.23,...,0.80,4313,18.90,455.70,5,233.47,41.95,Offer 10% discount,Excess inventory detected.,normal


##Adjusting for farm input and operations cost to account for geopolitical instability


In [14]:
def oil_shock_multiplier(scenario):

    if scenario == "normal":
        return np.random.uniform(0.98, 1.03)

    elif scenario == "moderate_disruption":
        return np.random.uniform(1.08, 1.18)

    elif scenario == "severe_disruption":
        return np.random.uniform(1.20, 1.40)

    else:
        return np.random.uniform(1.45, 1.80)

df["oil_price_shock_multiplier"] = df[
    "hormuz_scenario"
df["adjusted_fertilizer_index"] = (
].apply(oil_shock_multiplier)

In [15]:
df["adjusted_diesel_price"] = (
    df["diesel_price"]
    * df["oil_price_shock_multiplier"]
).round(2) #adjust oil prices to account of hormuz closure or supply shock

In [17]:
df["adjusted_fertilizer_index"] = (
    df["fertilizer_market_index"] * ( 1 + ((df["oil_price_shock_multiplier"] - 1)
            * 0.75 ))).round(2)#adjust fertilizer prices to account of hormuz closure or supply shock

In [18]:
df["adjusted_logistics_cost"] = ( df["logistics_cost"] * ( 1 + ((df["oil_price_shock_multiplier"] - 1)
            * 0.60))).round(2) #adjust logistics prices to account of hormuz closure or supply shock

In [20]:
df["farm_input_inflation_pct"] = (
    ((df["adjusted_fertilizer_index"] - df["fertilizer_market_index"])/ df["fertilizer_market_index"]
    )* 100).round(2) #calculate percentage inflation due to hormuz closure or supply shock

In [26]:
df["revenue_at_risk"] = (
    df["revenue"] * (
        df["farm_input_inflation_pct"] / 100)).round(2) #calculate revenue at risk or variability in revenue due to external factor

In [24]:
# ---------------------------------------------------------
# Adjusted unit production cost after geopolitical shock
# ---------------------------------------------------------

df["adjusted_unit_cost"] = (
    df["unit_cost"]
    * (1 + (df["farm_input_inflation_pct"] / 100)* 0.55)).round(2)

In [25]:
df["margin_compression_pct"] = (
    (
        (df["adjusted_unit_cost"] - df["unit_cost"] )/ df["unit_cost"] )
    * 100
).round(2)#the percentage increase in operational costs caused by inflation, geopolitical disruption, or supply-chain pressure.
#how much profitability is being squeezed
#costs are rising faster than revenue.

#This reduces:

#profitability
#operating margin
#business sustainability

#simulate different geopolitical strategy

In [26]:
def geopolitical_strategy(row):

    if row["hormuz_scenario"] == "closure":
        return "Increase prices aggressively"

    elif row["farm_input_inflation_pct"] > 15:
        return "Adjust pricing to protect margins"

    elif row["inventory_level"] > 4000:
        return "Offer targeted discounts"

    else:
        return "Maintain current pricing"

df["geopolitical_pricing_strategy"] = df.apply(
    geopolitical_strategy,
    axis=1
)

In [27]:
def geopolitical_summary(row):

    if row["hormuz_scenario"] == "closure":

        return (
            "Severe geopolitical disruption detected. "
            "Fuel and fertilizer costs are rising sharply."
        )

    elif row["farm_input_inflation_pct"] > 10:

        return (
            "Input cost inflation increasing operational risk."
        )

    else:

        return (
            "Market conditions remain stable."
        )

df["geopolitical_summary"] = df.apply(
    geopolitical_summary,
    axis=1
)

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go

import warnings
warnings.filterwarnings("ignore")

#Clean the data

In [4]:
df = pd.read_csv("/content/sample_data/AI_Smart_Agribusiness_Final_3.csv")

In [14]:
df.head(3)

,farm_id,region,crop_type,soil_moisture_%,soil_pH,temperature_C,rainfall_mm,humidity_%,sunlight_hours,irrigation_type,...,hormuz_scenario,oil_price_shock_multiplier,adjusted_diesel_price,adjusted_fertilizer_index,adjusted_logistics_cost,farm_input_inflation_pct,adjusted_unit_cost,margin_compression_pct,geopolitical_pricing_strategy,geopolitical_summary
0,FARM0001,North India,Wheat,35.95,5.99,17.79,75.62,77.03,7.27,NaN,...,normal,1.007821,4.44,131.76,507.18,0.59,305.48,0.33,Maintain current pricing,Market conditions remain stable.
1,FARM0002,South USA,Soybean,19.74,7.24,30.18,89.91,61.13,5.67,Sprinkler,...,severe_disruption,1.303522,5.06,152.07,294.32,22.77,411.57,12.52,Adjust pricing to protect margins,Input cost inflation increasing operational risk.
2,FARM0003,South USA,Wheat,29.32,7.16,27.37,265.43,68.87,8.23,Drip,...,normal,1.022939,3.86,112.80,461.97,1.72,235.68,0.95,Offer targeted discounts,Market conditions remain stable.


#Data Cleaning.....

In [12]:
df.drop(columns=['Unnamed: 0.1', 'Unnamed: 0', 'sensor_id'], inplace= True) #drop unnessary columns

In [21]:
from sklearn.impute import SimpleImputer

num_col = df.select_dtypes(include = 'number').columns
cat_col = df.select_dtypes(include = ['object','category']).columns

#fill missing numerical columns with the mean
num_imputer = SimpleImputer(strategy='mean')
df[num_col] = num_imputer.fit_transform(df[num_col])

#fill missing categorical with mode
col_imputer = SimpleImputer(strategy = 'most_frequent')
df[cat_col] = col_imputer.fit_transform(df[cat_col])

print(df.isnull().sum())

farm_id                          0
region                           0
crop_type                        0
soil_moisture_%                  0
soil_pH                          0
                                ..
farm_input_inflation_pct         0
adjusted_unit_cost               0
margin_compression_pct           0
geopolitical_pricing_strategy    0
geopolitical_summary             0
Length: 65, dtype: int64


#DESCRIBTIVE STATISTICS

In [27]:
total_revenue = df["revenue"].sum()

avg_profit_margin = df["profit_margin"].mean()

avg_revenue_risk = df["revenue_at_risk"].mean()

avg_demand = df["demand_index"].mean()

print("Total Revenue:", total_revenue)
print("Average Profit Margin:", avg_profit_margin)
print("Average Revenue Risk:", avg_revenue_risk)
print("Average Demand Index:", avg_demand)

Total Revenue: 890228676.0884343
Average Profit Margin: 32.83398
Average Revenue Risk: 128507.05478
Average Demand Index: 0.9836199999999999


*Total Revenue: 890,228,676.09
This is the sum of all revenue generated across the dataset.
It means the business generated approximately $890.23 million in sales.

*Average Profit Margin: 32.83%
Profit margin measures how much profit is retained from revenue after costs.
An average margin of 32.83% means: For every $1 of revenue, the business keeps about $0.33 as profit.

*Average Revenue Risk: 128,507.05
This likely represents the average financial exposure or uncertainty in revenue.
Depending on how you calculated it, it may reflect:
*   volatility,
*   potential losses,
*   forecast uncertainty,
*   supply chain disruptions,
*   climate/oil-price shocks,
*   or demand instability.

Interpretation:
On average, about $128.5k of revenue is exposed to risk or variability per observation/unit/region/time period.
If this came from simulation or shock modeling, it indicates how sensitive revenue is to disruptions.

*Average Demand Index: 0.9836
A demand index is usually a normalized indicator where:

*   1.0 = baseline/expected demand
*   greater 1.0 = above-normal demand
*   <1.0 = below-normal demand


#summary of the statistics above:
Together, these metrics tell a useful business story:

Revenue is very high (~$890M)
Profitability is healthy (~33% margin)
There is moderate revenue uncertainty/risk
Demand remains close to stable baseline levels

This combination suggests:

the business is profitable,
but external shocks or operational volatility may still threaten revenue stability.

#estimate revenue by crop tpe

In [35]:
crop_revenue = df.groupby('crop_type')['revenue'].sum().reset_index()

fig = px.bar(crop_revenue,
             x='crop_type',
             y = 'revenue',
             title = 'Revenue by crop type',
             text_auto= True)
# Make ALL chart text bold
fig.update_layout(
    title={
        'x': 0.5,
        'xanchor': 'center',
        'font': dict(
            family='Arial Black',
            size=22
        )
    },

    font=dict(
        family='Arial Black',
        size=14
    ),

    xaxis=dict(
        title='<b>Crop Type</b>',
        tickfont=dict(size=13)
    ),

    yaxis=dict(
        title='<b>Revenue</b>',
        tickfont=dict(size=13)))

# Make bar labels bold
fig.update_traces(
    #texttemplate='<b>%{text}</b>',
    textposition='outside',
    textfont=dict(
        family='Arial Black',
        size=12))
fig.show()

*Crops dominating revenue:

  Soybean generates the highest revenue, followed closely by maize and cotton. These crops are the primary revenue drivers in the portfolio, meaning they contribute the largest share of overall income and likely have the strongest combination of demand, yield, and market price performance. In contrast, rice and wheat contribute the least revenue, indicating lower profitability or weaker market performance relative to the top-performing crops.


*Where pricing focus should go:
     Pricing strategy should prioritize soybean, maize, and cotton because even small pricing improvements or margin increases in these high-revenue crops can produce significant overall financial impact. Dynamic pricing, premium positioning, or supply-chain efficiency efforts should be concentrated on these products first. Rice and wheat may require competitive or cost-sensitive pricing approaches instead of aggressive price increases, since their lower revenue contribution suggests weaker pricing power or lower margins.


*Which products deserve optimization:
     Soybean, maize, and cotton deserve the highest optimization priority because improving forecasting, logistics, production efficiency, inventory management, or market timing for these crops will likely generate the greatest return on investment. However, rice and wheat should also be reviewed strategically to determine whether:

     costs can be reduced,
     operational inefficiencies exist,
     demand can be improved,
     or whether resources should be reallocated toward higher-performing crops.

     

In short, optimization efforts should focus on maximizing profitability of the top-performing crops while evaluating whether low-performing products are worth restructuring, repositioning, or scaling down.

In [38]:
fig = px.bar(df,
             x = 'crop_type',
             y = 'profit_margin',
             color = 'crop_type',
             title = 'profit margin by crop type')
# Make ALL chart text bold
fig.update_layout(
    title={
        'x': 0.5,
        'xanchor': 'center',
        'font': dict(
            family='Arial Black',
            size=22
        )
    },

    font=dict(
        family='Arial Black',
        size=14
    ),

    xaxis=dict(
        title='<b>Crop Type</b>',
        tickfont=dict(size=13)
    ),

    yaxis=dict(
        title='<b>Profit margin</b>',
        tickfont=dict(size=13)))

# Make bar labels bold
fig.update_traces(
    #texttemplate='<b>%{text}</b>',
    textposition='outside',
    textfont=dict(
        family='Arial Black',
        size=12))
fig.show()

If maize has the highest profit margin, it changes the business interpretation significantly:


Revenue leader vs. profitability leader:
Even if soybean generates the highest total revenue, maize becomes the most financially efficient crop because it retains the largest percentage of profit after costs. This means maize may be delivering more earnings per dollar of sales.


Strategic pricing focus:
Maize should receive stronger pricing and market focus because it has greater room to generate profit without being heavily eroded by production or operational costs. Maintaining maize pricing power becomes critical since it contributes disproportionately to profitability.


Low-performing crops:
Since rice and wheat already have low revenue, comparing their margins becomes important. If they also have low margins, they may be underperforming operationally and could require:
cost reduction,
process redesign,
repositioning,
or reduced investment allocation.


Overall, maize having the highest profit margin suggests the company should prioritize profitability-centered strategies around maize while using soybean for scale and revenue stability.

#Crop yield prediction using remote sending: remote sensing → business value

In [49]:
fig = px.scatter(
    df,
    x="NDVI_index",
    y="yield_kg_per_hectare",
    color="crop_type",
    facet_col ="crop_type",
    facet_col_wrap = 3,
    title="NDVI vs Crop Yield",
    trendline="ols")
# Make ALL chart text bold
# Improve layout
fig.update_layout(
    height=800,
    width=1200,
    title_font=dict(size=22, family="Arial Black"),
    font=dict(size=14, family="Arial Black")
)

# Make subplot titles bold
fig.for_each_annotation(
    lambda a: a.update(text=f"<b>{a.text}</b>")
)

# Make axis titles bold
fig.update_xaxes(title_font=dict(size=18, family="Arial Black"))
fig.update_yaxes(title_font=dict(size=18, family="Arial Black"))

fig.show()

*  Business / Agricultural Implications

Cotton and maize appear most responsive to NDVI-based monitoring.
These crops may benefit most from remote sensing analytics and precision agriculture systems.

NDVI can likely support early yield forecasting and intervention planning.
Wheat and soybean show weak NDVI predictive power.
Additional variables should be incorporated into models:
*  rainfall,
*  soil moisture,
*  fertilizer application,
*  temperature stress,
*  disease pressure.

The large spread of points across all crops indicates:
yield is influenced by multiple interacting variables,
and NDVI alone is insufficient for high-accuracy prediction.

*  Key Takeaway

NDVI is a useful vegetation-health indicator, but its predictive strength varies by crop type. Cotton and maize exhibit clearer positive NDVI–yield relationships, while wheat and soybean yields appear less dependent on NDVI alone, suggesting the need for more advanced multi-factor yield models.